Draft: Space point display update -- 3dTo2d projection to Hit Association

Jjm321814 opened the pull request:
Changes to DrawSpacePoint.cxx, DrawSpacePoint.h  and spacepoint.py to prioritize use of space-point hit association object over space-points directly.

Main feature highlights are shown https://docs.google.com/presentation/d/13ozNkPdjZ_9fTyeltv52PiUiTGmpKgSa7UZUs5rXi50/edit#slide=id.p 

Space points direct code is removed, so we would need to guarantee that the association plot is always made.
Additionally the _dataByPlane array is filled by a custom HitFromSpacePoint object, rather than a shared object like larutil::Point2D, which may limit the generalizability of code to other systems. 

Occasional bugs have been observed with the shared highlighting feature that are fixed by left-clicking to adjust the view slightly.

Note the spacePoint ID displayed in the tool tip is just an index in the vector rather than a value saved to the spacePoint itself.

Commit:
First attempt at drawing space points with hit association

Commit:
Part way to a working version of the update

Commit:
Completed SpacePoint display update. Added QGraphicsItemGroup that links the space point hits across tpc views allowing for shared hover actions (color change + tool tip)

Commit:
forgot to remove print out message

marcodeltutto commented:
Thanks very much @Jjm321814 for your work on this! Looks very nice. My only comment is that this is effectively drawing hits, not spacepoints anymore. I'm afraid this can be a bit confusing, if the spacepoint reconstruction is not perfoerming well, there is no way to tell here because this is PR is bypassing the actual spacepoints and looking at the associated hits. 

I am wondering if we should make this optional?

Jjm321814 committed:
Moving changes to space point display to its own button

Jjm321814 committed:
Think I had a typo

Jjm321814 committed:
This may fix it?

Jjm321814 committed:
This may fix it?

Jjm321814 committed:
Finally should launch I think

Jjm321814 committed:
EVD renders but the gui is super spread out. Testing if its a box size or my new addition being wrong

Jjm321814 committed:
My module is breaking the ui a bit, so I updated a few rerefences in drawableItems

Jjm321814 committed:
Copying more link stuff

Jjm321814 committed:
Captialization issue

Jjm321814 committed:
Changed return of 3d_to_2d function in geometryHelper to give units of wire, time

Jjm321814 committed:
Tweaked units and view changes

Jjm321814 committed:
Debug

Jjm321814 committed:
Debug

Jjm321814 committed:
debug

Jjm321814 committed:
Changing which geometry helper is invoked

Jjm321814 committed:
debug

Jjm321814 committed:
Adjusting processing

Jjm321814 committed:
Restoring old units

Jjm321814 committed:
Changed way space points are drawn after adjustment to data extraction

Jjm321814 committed:
Minor tweak to tool tip

Jjm321814 committed:
Encoding final data return differently

Jjm321814 committed:
copy paste issue

Jjm321814 committed:
copy paste issue

Jjm321814 committed:
copy paste issue

Jjm321814 committed:
Okay finally something will work

Jjm321814 committed:
Okay finally something will work

Jjm321814 committed:
Tweaking highlight color

Jjm321814 committed:
Changes to when things are added to view

Jjm321814 committed:
Drawn object update

Jjm321814 committed:
Didn't work. Try drawing another way

Jjm321814 committed:
Okay now just need to make a hover parent

Jjm321814 committed:
Updated SpacePointID to come from object itself rather than index

Jjm321814 committed:
Adjusting debug print + error handling branch

Jjm321814 committed:
Update to space point ID caused an indexing bug, so adding debug

Jjm321814 committed:
Putting old indexing back

Jjm321814 committed:
debug

Jjm321814 committed:
debug

Jjm321814 committed:
Altering constructor call

Jjm321814 committed:
Revert + debug

Jjm321814 committed:
debug

Jjm321814 committed:
Fixing a bit of hardcoding

Jjm321814 committed:
Fixing a bit of hardcoding

Jjm321814 committed:
Fixing a bit of hardcoding

Jjm321814 committed:
Trying to address timing offsets

Jjm321814 committed:
Trying to address timing offsets

Jjm321814 committed:
Trying new idea

Jjm321814 committed:
Didn't work

Jjm321814 committed:
Adjusting order of shifts

Jjm321814 committed:
Checking unit conversion in python vs cxx. Should be the same

Jjm321814 committed:
Found one bug with tool tip

Jjm321814 committed:
Adjusting offset direction

Jjm321814 committed:
Didn't seem to work. Adding debug

Jjm321814 committed:
Trying adjustments on python side

Jjm321814 committed:
Treating both tpc the same

Jjm321814 committed:
One more try at standard processing

Jjm321814 committed:
One more try at standard processingA

Jjm321814 committed:
One more try at standard processingA

Jjm321814 committed:
Mised a save

Jjm321814 committed:
Changed tool tip to be simpler for me to use for debug

Jjm321814 committed:
adding debug

Jjm321814 committed:
more debug

Jjm321814 committed:
more debug

Jjm321814 committed:
changing loop method

Jjm321814 committed:
Debug

Jjm321814 committed:
Different views have drawstically different timestamps

Jjm321814 committed:
Adding a plane readout as a guess at cause

Jjm321814 committed:
Found one bug

Jjm321814 committed:
Removing one of the offset corrections

Jjm321814 committed:
reverted to the ancient logic

Jjm321814 committed:
Nearly got it! Updating tool tip

Jjm321814 committed:
Forgot a unit conversion

Jjm321814 committed:
Removing lots of debug

Jjm321814 commented:
Updated slides 11-18 [here](https://docs.google.com/presentation/d/13ozNkPdjZ_9fTyeltv52PiUiTGmpKgSa7UZUs5rXi50/edit#slide=id.g341e7a1fb45_0_6)
I agree with making the hit association display optional, and I have split the two space point displays into two separate drop down buttons in the TPC view. 
This splitting necessitated a new DrawSpacePointHits.cxx class as well as SpacePointHits.py module to be added. These are basically copy-pastes of the proposed update at the previous step with class names changed.

Making the old space point display function well involved updates to SimpleGeometryHelper.cxx (mostly helping it to figure out what plane ID =4 means [TPC 1 plane 0]) and adding that offset into appropriate calls in Point_3Dto2D. 
I modified the object saved for spacePoint from a lar:util:Point2d to a custom object with a space point ID, time, wire, plane number attribute. 
This makes it more similar to the DrawSpacePointHits.cxx object so I can recycle some of the tool tip + highlighting code between the two .py modules. 
SpacePoint.py features a small hardcoded offset of 13 (time units) [line 99] that may be unnecessary and is easy to eliminate. 

Additional potentially undesirable features are highlighted in the linked slides 17 and 18, both mostly stemming from analyzing every space point to display on each plane. Since spacePoints lack a tpc entry I don't think these features are avoidable.

Jjm321814 committed:
fixing merge conflict

Jjm321814 committed:
fixed function definition

Jjm321814 committed:
updating wire readout construction

Jjm321814 committed:
Try again

Jjm321814 committed:
Try again

Jjm321814 committed:
Try again

Jjm321814 committed:
Try again

Jjm321814 committed:
Converted unique pointer. I am worried about how I am handling the delete call though

Jjm321814 committed:
typo

Jjm321814 committed:
Trying a different shceme to initialize readout

Jjm321814 committed:
Wrong Argument type for fcl file pass

Jjm321814 committed:
Wrong template type

Jjm321814 committed:
adding types for template:

Jjm321814 committed:
More template argument adjustment

Jjm321814 committed:
More template argument adjustment

Jjm321814 committed:
More template argument adjustment

Jjm321814 committed:
More template argument adjustment

Jjm321814 committed:
Hey we made progress

Jjm321814 committed:
copy paste issue

Jjm321814 committed:
swap geo calls for wire readout

Jjm321814 committed:
Swap geo calls and update error message

Jjm321814 committed:
Adding interface for new geometry object

Jjm321814 committed:
wrong include

Jjm321814 committed:
wrong include

Jjm321814 committed:
typo

Jjm321814 committed:
Update raw base constructor

Jjm321814 committed:
Updated draw wires in the same way as draw raw digit

Jjm321814 committed:
Raw base geometry calls update

Jjm321814 committed:
Guess that FEB will work like this

Jjm321814 committed:
Fix to FEB

Jjm321814 committed:
Fixed typo

Jjm321814 committed:
Normal update for this file

Jjm321814 committed:
typo

Jjm321814 committed:
RecoViewer standard update times

Jjm321814 committed:
Wrong recobase construction

Jjm321814 committed:
Standard updates

Jjm321814 committed:
Standard updates for DrawHit

Jjm321814 committed:
Missed one

Jjm321814 committed:
Standard updates for DrawMCTrack + simpleGeoHelper update

Jjm321814 committed:
Standard DrawMCTruth updates

Jjm321814 committed:
Standard updates DrawNumuSelection

Jjm321814 committed:
Making this compile. New function may not take op channel though?

Jjm321814 committed:
Standard updates

Jjm321814 committed:
Standard updates

Jjm321814 committed:
Standard updates

Jjm321814 committed:
Standard updates

Jjm321814 committed:
standard update. Last one I think?

Jjm321814 committed:
nope missed one

Jjm321814 committed:
larV10 merge conflict

Jjm321814 committed:
Messed up on the merge in a spot

Jjm321814 committed:
Merge issue

Jjm321814 committed:
Other merge issue

Jjm321814 committed:
Trying to add wire readout interface. May have wrong service manager tag

Jjm321814 committed:
Maybe the old label was wrong?

Jjm321814 committed:
Change to all wire readout

Jjm321814 committed:
Debug

Jjm321814 committed:
Added debug

Jjm321814 committed:
Adjusted debug

Jjm321814 committed:
Adjusted debug

Jjm321814 committed:
Adjusted debug

Jjm321814 committed:
Adjusted debug

Jjm321814 committed:
Adjusted debug

Jjm321814 committed:
Adjusted debug

Jjm321814 committed:
Adjusted wire readout calls

Jjm321814 committed:
Realized we had doubled up some fixes

Jjm321814 committed:
Missed and arg to delete

Jjm321814 committed:
Missed and arg to delete

Jjm321814 committed:
Forgot to call function

Jjm321814 committed:
Adding appropriate calls from python to cxx

Jjm321814 committed:
Missed some initialization

Jjm321814 committed:
Update to crt module aux det

Jjm321814 committed:
Typo

Jjm321814 committed:
Typo

Jjm321814 committed:
Typo

Jjm321814 committed:
Merge branch 'master' into feature/JJM_SpaceChargeTweaks

Jjm321814 commented:
Just repinging that this is updated to be compatible with LArSoft V10, and there are two complimentary space point displays for both hits and the 3d space point itself.

Jjm321814 committed:
Updating blips to show pdg. Please don't PR this one in

Jjm321814 committed:
Guess at library

Jjm321814 committed:
Guess at library

Jjm321814 committed:
Guess at library

Jjm321814 committed:
Guess at library

Jjm321814 committed:
Guess at library

Commit:
hardcoded library instead

Commit:
one fix of cmake. AGAIN DONT PR THIS

Jjm321814 committed:
Give me a handle not a vector

Commit:
fixed compile bugs. DONT PR

Jjm321814 committed:
Fixed a typo

Jjm321814 committed:
python stuff

Jjm321814 committed:
Whoops wrong condition
